### PR TITLE
Unit 10 - HW, part 3

### PR DESCRIPTION
**Pull request by AlinaD19**
Unit 10 - HW, part 3

### CODE REVIEW

File: src/pages/CheckoutSecond.page.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    get checkoutItems() { return this.page.locator('.cart_item'); }
+
+    get itemsTotal() { return this.page.locator('.summary_subtotal_label'); }
+
+    get itemsTax() { return this.page.locator('.summary_tax_label'); }
+
+    get orderTotalPrice() { return this.page.locator('.summary_total_label'); }
+
+    async goToFinishCheckoutFlow() {
+        await this.finishButton.click();
```
Comment (hlazkov) [inline src/pages/CheckoutSecond.page.js:25]: I wouldn't wrap one liner with other one liner, since it adds not a lot of sence. But here we have descriptive name, so IDK :)

File: src/pages/CheckoutSecond.page.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    async extractNumberFromStr(str) {
```
Comment (hlazkov) [inline src/pages/CheckoutSecond.page.js]: 1. This method should be definitely moved to utils, its not a part of the page.
2. Its not async! I've checked!

File: src/pages/CheckoutSecond.page.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        const match = str.match(/\$([\d.]+)/);
+        let number;
+        if (match) {
+            const numberString = match[1];
+            number = parseFloat(numberString);
+        }
+        return number;
+    }
+
+    /* async getNumberOfItemsOnCheckout() {
```
Comment (hlazkov) [inline src/pages/CheckoutSecond.page.js]: Lets clean this up

File: src/pages/ShopingCart.page.js [hunk 4]
```diff
@@ -40,4 +42,32 @@ export class ShopingCartPage extends BaseSwagLabPage {
...
+    }
+
+    async collectedItems() {
+        const names = await this.getAllCartItemsName();
+        const prices = await this.getAllCartItemsPrice();
+        const descriptions = await this.gettAllCartItemsDescription();
+
+        const allItemsInformation = [];
+
+        // Assuming names, prices, and descriptions arrays have the same length
```
Comment (hlazkov) [inline src/pages/ShopingCart.page.js]: <3

Comment (hlazkov) [reply src/pages/ShopingCart.page.js]: And indexes are the same xD

File: src/tests/ckeckout.spec.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        await loginPage.performLogin(login, password);
+
+        await expect(inventoryPage.headerTitle).toBeVisible();
+    });
+
+    test('Proceed to Checkout, place an Order and verify products', async ({
+        inventoryPage, shopingCartPage, checkoutFirstPage, checkoutSecondPage,
+    }) => {
+        // Perform adding random quantity of random items to the Shopping Cart
+        const arrayOfRandomIndex = await inventoryPage.arrayOfRandomItemIndex();
```
Comment (hlazkov) [inline src/tests/ckeckout.spec.js]: oh please rewrite arrayOfRandomItemIndex into the plain code, Im not sure its reusable at all.

File: src/tests/ckeckout.spec.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        const unicRandomIndex = [...new Set(arrayOfRandomIndex)];
+        let itemsOnPageAfterAdding = await inventoryPage.getNumberOfItems();
+
+        for (let i = 0; i < unicRandomIndex.length; i += 1) {
+            if (unicRandomIndex[i] < itemsOnPageAfterAdding) {
+                await inventoryPage.addItemToCartById(unicRandomIndex[i]);
+                itemsOnPageAfterAdding -= 1;
```
Comment (hlazkov) [inline src/tests/ckeckout.spec.js]: Do we add ALL items in random order? Its hard to follow the logic here :(

File: src/tests/ckeckout.spec.js (hunk 5 above)
Comment (hlazkov) [reply src/tests/ckeckout.spec.js]: But you reuse it! Oh, please don't do that! As far as I got that code, we create and array with length of items on the page cont and fill it with numbers 0-5. Then we use these numbers as indexes but in case number of items on the page > 5, we will get an array with the first 6 elements and thats it. Then we add it in random order but its not the random items, there are still first items to be added.

**Review by hlazkov — CHANGES_REQUESTED** [review]
The biggest issue I see here is with the random items logic, but please check on other comments as well.

File: src/pages/CheckoutSecond.page.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+const { ShopingCartPage } = require('./ShopingCart.page');
+
+export class CheckoutSecondPage extends ShopingCartPage {
```
Comment (AlinaZhykharieva) [inline src/pages/CheckoutSecond.page.js]: are we sure we want to extend ShopingCartPage?not Base Page?

File: src/tests/ckeckout.spec.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
+const { expect } = require('@playwright/test');
+const { test } = require('../fixture');
+const { login, password } = require('../credentials/credentials.json');
+const { extractNumberFromStr } = require('../utils/methods.page');
+const { arrayOfRandomItemIndex } = require('../utils/methods.page');
```
Comment (AlinaZhykharieva) [inline src/tests/ckeckout.spec.js]: can be combined to one line

File: src/utils/methods.page.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+// This function is created to exctract only numbers from using string
```
Comment (AlinaZhykharieva) [inline src/utils/methods.page.js]: We can not name this module with page extension, this is not the page

File: src/tests/ckeckout.spec.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        const allItemInfoInCart = await shopingCartPage.collectedItems();
+
+        // Go to First Checkout step and verify page
+        await shopingCartPage.goToCheckoutStep();
+        await expect(checkoutFirstPage.headerTitle).toBeVisible();
+
+        // Fill in all required fields
+        await checkoutFirstPage.firstNameField.fill('Rayan');
+        await checkoutFirstPage.lastNameField.fill('Gosling');
+        await checkoutFirstPage.zipCodeField.fill('12345');
```
Comment (AlinaZhykharieva) [inline src/tests/ckeckout.spec.js]: for better management of the dest data, I will propose to keep Rayan in a separate module

**Review by AlinaZhykharieva — CHANGES_REQUESTED** [review]

File: src/tests/checkout.spec.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        // Check if all products were added to the Shopping Cart
+        const addedItemsInCart = await inventoryPage.getNumberOfItemsInCart();
+        await expect(inventoryPage.shopingCartBadge).toHaveText(addedItemsInCart);
+
+        // Go to Shopping Cart and verify page
+        await inventoryPage.shopingCart.click();
+        await expect(shopingCartPage.headerTitle).toBeVisible();
+
+        // Collect information of added items in shopping cart to verify details on Checout later
+        const allItemInfoInCart = await shopingCartPage.collectedItems();
```
Comment (ovartem) [inline src/tests/checkout.spec.js:32]: for consistency, I'd collect items from inventory page (only those I'm going to add to cart) and add additional comparison checking those items are in cart and then on checkout

**Review by ovartem — APPROVED** [review]
Well done!
I've approved the PR but if you want - one comment was left and can be fixed.
Great Job!
